### DISCUSSION TITLE
Using VDD as the ADC reference in STM32 not possible? (adc_stm32.c)

### DISCUSSION THREAD

**Discussion by flabou** [zephyrproject-rtos/zephyr#64905] [Q&A]
Is it possible to use VDD as a reference for the ADC? I tried and got the error "Invalid channel reference" which I found comes from the following snippet in "adc_stm32.c" driver:

```
if (channel_cfg->reference != ADC_REF_INTERNAL) {
        LOG_ERR("Invalid channel reference");
        return -EINVAL;
}
```

This suggests to me that it isn't possible to use VDD as reference for the ADC, is this correct?

It is a drawback for me. Although, generally, it is better to use internal reference than VDD for accuracy, here is a case (my case) when it is more accurate to use VDD as reference:

VDD supplies a divider bridge to measure the value of a thermistor. Since the divider bridge is supplied by VDD and not internal voltage reference, using VDD as a reference allows to measure the ratio between VDD and the bottom resistor, without caring about the accuracy of VDD. Using internal reference adds the error of VDD on top of all the other errors.

Additionally, I think this also means it is not possible to use the external references, which could potentially be more accurate than the internal one.

**Comment by gautierg-st** [comment]
This `channel_cfg->reference` is unused in the STM32 ADC driver. In the end, the reference value is whatever is provided on VDDA/VREF+/VREF- pins, and the corresponding value of the `vref_mv` property in the ADC node in the dts. Usually, it is indeed VDD that is connected to the previously mentioned pins and `vref_mv` default value is 3300 (mV).
So I think your program should work well with the ADC_REF_INTERNAL value.

**Reply by flabou** [reply]
So, in the end it seems to work, but I think it is confusing that my reference has to be configured to `zephyr,reference  ="ADC_REF_INTERNAL"`, I would have assumed it should be configured to `zephyr,reference  ="ADC_REF_VDD_1"`.

**Comment by ypearson-bdai** [comment]
ADC_REF_VDD_1 doesn't seem to work here too
